index.js

```diff
@@ -1,80 +1,106 @@
-		useRelativePath: false,
-		name: "[hash].[ext]"
-	};
+  var query = loaderUtils.getOptions(this) || {};
+  var configKey = query.config || "fileLoader";
+  var options = this.options[configKey] || {};
+  var config = {
+    publicPath: false,
+    useRelativePath: false,
+    name: "[hash].[ext]"
+  };
 
-	// options takes precedence over config
-	Object.keys(options).forEach(function(attr) {
-		config[attr] = options[attr];
-	});
+  // options takes precedence over config
+  Object.keys(options).forEach(function(attr) {
+    config[attr] = options[attr];
+  });
 
-	// query takes precedence over config and options
-	Object.keys(query).forEach(function(attr) {
-		config[attr] = query[attr];
-	});
+  // query takes precedence over config and options
+  Object.keys(query).forEach(function(attr) {
+    config[attr] = query[attr];
+  });
 
-	var context = config.context || this.options.context;
-	var url = loaderUtils.interpolateName(this, config.name, {
-		context: context,
-		content: content,
-		regExp: config.regExp
-	});
+  var context = config.context || this.options.context;
+  var url = loaderUtils.interpolateName(this, config.name, {
+    context: context,
+    content: content,
+    regExp: config.regExp
+  });
 
-	var outputPath = "";
+  var outputPath = "";
+  if (config.outputPath) {
+    // support functions as outputPath to generate them dynamically
+    outputPath = (
+      typeof config.outputPath === "function"
+      ? config.outputPath(url)
+      : config.outputPath + url
+    );
+  }
 
-	var filePath = this.resourcePath;
-	if (config.useRelativePath) {
-		var issuerContext = this._module && this._module.issuer && this._module.issuer.context || context;
-		var relativeUrl = issuerContext && path.relative(issuerContext, filePath).split(path.sep).join("/");
-		var relativePath = relativeUrl && path.dirname(relativeUrl) + "/";
-		if (~relativePath.indexOf("../")) {
-			outputPath = path.posix.join(outputPath, relativePath, url);
-		} else {
-			outputPath = relativePath + url;
-		}
-		url = relativePath + url;
-	} else if (config.outputPath) {
-		// support functions as outputPath to generate them dynamically
-		outputPath = (
-			typeof config.outputPath === "function"
-			? config.outputPath(url)
-			: config.outputPath + url
-		);
-		url = outputPath;
-	} else {
-		outputPath = url;
-	}
+  if (config.useRelativePath) {
+    // Only the dirname is needed in this case.
+    outputPath = outputPath.replace(url, "");
 
-	var publicPath = "__webpack_public_path__ + " + JSON.stringify(url);
-	if (config.publicPath) {
-		// support functions as publicPath to generate them dynamically
-		publicPath = JSON.stringify(
-			typeof config.publicPath === "function"
-			? config.publicPath(url)
-			: config.publicPath + url
-		);
-	}
+    // We have access only to entry point relationships. So we work with this relations.
+    var issuerContext = (this._module && this._module.issuer && this._module.issuer.context) || context;
```

Comment:
In the gist a newline `\n` after the follwing lines and I'm 👍 on Code Style 😛 L50, 65, 74, 84